### ISSUE TITLE
.pt vs .engine inference results differ even with same input (YOLOv11 segmentation, no errors)

### ISSUE THREAD

**Issue by myungki-park** [ultralytics/ultralytics#21033]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Search before asking

- [x] I have searched the HUB [issues](https://github.com/ultralytics/hub/issues) and [discussions](https://github.com/ultralytics/hub/discussions) and found no similar questions.

### Question

I'm observing a discrepancy between .pt and .engine inference results using the same image and model weights in YOLOv11 segmentation. There is no error, but the outputs differ significantly (boxes, confidences, or masks), even when using the same Ultralytics inference API.

System Info:

Ultralytics version: 8.3.128

Torch version: 2.3.1+cu121

TensorRT version: 8.6.0

OS: Linux

GPU: A100

CUDA : 12.2

belows are the .pt -> .engine coversion and comparison script in python, 

from ultralytics import YOLO
import numpy as np
import cv2
import os

# ======== CONFIG ========
PT_MODEL_PATH = r"/home/jovyan/project/APP_PCS_ETX/weights/tray_detector.pt"
ENGINE_MODEL_PATH = r"/home/jovyan/project/APP_PCS_ETX/weights/tray_detector.engine"
TEST_IMAGE_PATH = r"/home/jovyan/project/APP_PCS_ETX/testdata/20250513/20250513012211901^C426A^LDETX^M6242P^82F_10X11_512P^PDA9790774.jpg"
IMGSZ = 640  # Must match training and export size
TOLERANCE = 1e-3  # acceptable numerical diff
# ========================

def export_to_engine(pt_path, engine_path, imgsz):
    print(f"[INFO] Exporting {pt_path} to TensorRT...")
    model = YOLO(pt_path)
    model.export(format="engine", imgsz=imgsz, dynamic=False, simplify=True)
    assert os.path.exists(engine_path), "Export failed — .engine file not found"
    print("[INFO] Export complete.")

def load_image(path):
    img = cv2.imread(path)
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    return img

def summarize_result(r):
    boxes = r.boxes.xywhn.cpu().numpy() if r.boxes else None
    confs = r.boxes.conf.cpu().numpy() if r.boxes else None
    classes = r.boxes.cls.cpu().numpy() if r.boxes else None
    masks = r.masks.data.cpu().numpy() if r.masks else None
    return boxes, confs, classes, masks

def compare_np(a, b, name, tol=1e-3):
    if a is None or b is None:
        print(f"[WARN] {name} - One is None → Skipping")
        return
    if a.shape != b.shape:
        print(f"[FAIL] {name} - Shape mismatch: {a.shape} vs {b.shape}")
        return
    diff = np.abs(a - b)
    max_diff = np.max(diff)
    print(f"[INFO] {name} - Max Diff: {max_diff:.5f}")
    if max_diff < tol:
        print(f"[PASS] {name} Equivalent")
    else:
        print(f"[FAIL] {name}  NOT Equivalent")

def run_comparison(pt_model_path, engine_model_path, image_path, imgsz):
    print("[INFO] Loading models...")
    model_pt = YOLO(pt_model_path)
    model_trt = YOLO(engine_model_path)

    img = load_image(image_path)

    print("[INFO] Running inference...")
    result_pt = model_pt(img, imgsz=imgsz, stream=False, verbose=False)[0]
    result_trt = model_trt(img, imgsz=imgsz, stream=False, verbose=False)[0]

    print("[INFO] Summarizing results...")
    boxes_pt, confs_pt, classes_pt, masks_pt = summarize_result(result_pt)
    boxes_trt, confs_trt, classes_trt, masks_trt = summarize_result(result_trt)

    print("\n===== Comparison Results =====")
    compare_np(boxes_pt, boxes_trt, "Boxes", tol=TOLERANCE)
    compare_np(confs_pt, confs_trt, "Confs", tol=TOLERANCE)
    compare_np(classes_pt, classes_trt, "Classes", tol=0)  # Class indices must match exactly
    compare_np(masks_pt, masks_trt, "Masks", tol=TOLERANCE)

    result_pt.save(filename="output_pt.jpg")
    result_trt.save(filename="output_trt.jpg")
    print("[INFO] Saved visualization images: output_pt.jpg, output_trt.jpg")

if __name__ == "__main__":
    if not os.path.exists(ENGINE_MODEL_PATH):
        export_to_engine(PT_MODEL_PATH, ENGINE_MODEL_PATH, IMGSZ)
    run_comparison(PT_MODEL_PATH, ENGINE_MODEL_PATH, TEST_IMAGE_PATH, IMGSZ)

**And Belows are the results of the above script(Sorry for pasting the text, but i can't upload imgeas due to company's policy)**

⬢ [notebook] ❯  /usr/bin/env /home/jovyan/.conda/envs/yolo/bin/python /home/jovyan/.local/share/code-server/extensions/ms-python.debugpy-2025.6.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher 41683 -- /home/jovyan/project/APP_PCS_ETX/code/tooling/convert_pt_to_tensorrt.py 
[INFO] Exporting /home/jovyan/project/APP_PCS_ETX/weights/tray_detector.pt to TensorRT...
WARNING ⚠️ TensorRT requires GPU export, automatically assigning device=0
Ultralytics 8.3.128 🚀 Python-3.10.14 torch-2.3.1+cu121 CUDA:0 (NVIDIA A100 80GB PCIe MIG 2g.20gb, 19968MiB)
YOLO11n-seg summary (fused): 113 layers, 2,834,763 parameters, 0 gradients, 10.2 GFLOPs

PyTorch: starting from '/home/jovyan/project/APP_PCS_ETX/weights/tray_detector.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 37, 8400), (1, 32, 160, 160)) (6.1 MB)

ONNX: starting export with onnx 1.17.0 opset 17...
ONNX: slimming with onnxslim 0.1.52...
ONNX: export success ✅ 2.5s, saved as '/home/jovyan/project/APP_PCS_ETX/weights/tray_detector.onnx' (11.1 MB)

TensorRT: starting export with TensorRT 8.6.0...
[06/12/2025-16:38:43] [TRT] [I] [MemUsageChange] Init CUDA: CPU +2, GPU +0, now: CPU 728, GPU 567 (MiB)
[06/12/2025-16:38:48] [TRT] [I] [MemUsageChange] Init builder kernel library: CPU +1625, GPU +302, now: CPU 2429, GPU 869 (MiB)
[06/12/2025-16:38:48] [TRT] [I] ----------------------------------------------------------------
[06/12/2025-16:38:48] [TRT] [I] Input filename:   /home/jovyan/project/APP_PCS_ETX/weights/tray_detector.onnx
[06/12/2025-16:38:48] [TRT] [I] ONNX IR version:  0.0.8
[06/12/2025-16:38:48] [TRT] [I] Opset version:    17
[06/12/2025-16:38:48] [TRT] [I] Producer name:    pytorch
[06/12/2025-16:38:48] [TRT] [I] Producer version: 2.3.1
[06/12/2025-16:38:48] [TRT] [I] Domain:           
[06/12/2025-16:38:48] [TRT] [I] Model version:    0
[06/12/2025-16:38:48] [TRT] [I] Doc string:       
[06/12/2025-16:38:48] [TRT] [I] ----------------------------------------------------------------
[06/12/2025-16:38:48] [TRT] [W] onnx2trt_utils.cpp:374: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
TensorRT: input "images" with shape(1, 3, 640, 640) DataType.FLOAT
TensorRT: output "output0" with shape(1, 37, 8400) DataType.FLOAT
TensorRT: output "output1" with shape(1, 32, 160, 160) DataType.FLOAT
TensorRT: building FP32 engine as /home/jovyan/project/APP_PCS_ETX/weights/tray_detector.engine
[06/12/2025-16:38:48] [TRT] [I] Graph optimization time: 0.0320165 seconds.
[06/12/2025-16:38:48] [TRT] [I] Local timing cache in use. Profiling results in this builder pass will not be stored.
[06/12/2025-16:40:47] [TRT] [I] Detected 1 inputs and 5 output network tensors.
[06/12/2025-16:40:47] [TRT] [I] Total Host Persistent Memory: 592608
[06/12/2025-16:40:47] [TRT] [I] Total Device Persistent Memory: 103936
[06/12/2025-16:40:47] [TRT] [I] Total Scratch Memory: 0
[06/12/2025-16:40:47] [TRT] [I] [MemUsageStats] Peak memory usage of TRT CPU/GPU memory allocators: CPU 2 MiB, GPU 133 MiB
[06/12/2025-16:40:47] [TRT] [I] [BlockAssignment] Started assigning block shifts. This will take 228 steps to complete.
[06/12/2025-16:40:47] [TRT] [I] [BlockAssignment] Algorithm ShiftNTopDown took 16.4986ms to assign 10 blocks to 228 nodes requiring 22553600 bytes.
[06/12/2025-16:40:47] [TRT] [I] Total Activation Memory: 22553600
[06/12/2025-16:40:47] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in building engine: CPU +2, GPU +11, now: CPU 2, GPU 11 (MiB)
TensorRT: export success ✅ 127.5s, saved as '/home/jovyan/project/APP_PCS_ETX/weights/tray_detector.engine' (14.7 MB)

Export complete (128.4s)
Results saved to /home/jovyan/project/APP_PCS_ETX/weights
Predict:         yolo predict task=segment model=/home/jovyan/project/APP_PCS_ETX/weights/tray_detector.engine imgsz=640  
Validate:        yolo val task=segment model=/home/jovyan/project/APP_PCS_ETX/weights/tray_detector.engine imgsz=640 data=/home/jovyan/yolo/data/TrayDetection/YOLODataset/dataset.yaml  
Visualize:       https://netron.app
[INFO] Export complete.
[INFO] Loading models...
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
[INFO] Running inference...
Loading /home/jovyan/project/APP_PCS_ETX/weights/tray_detector.engine for TensorRT inference...
[06/12/2025-16:40:49] [TRT] [I] The logger passed into createInferRuntime differs from one already provided for an existing builder, runtime, or refitter. Uses of the global logger, returned by nvinfer1::getLogger(), will return the existing value.
[06/12/2025-16:40:49] [TRT] [I] Loaded engine size: 14 MiB
[06/12/2025-16:40:49] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in engine deserialization: CPU +0, GPU +11, now: CPU 0, GPU 11 (MiB)
[06/12/2025-16:40:49] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +21, now: CPU 0, GPU 32 (MiB)
[INFO] Summarizing results...

===== 🔍 Comparison Results =====
[FAIL] Boxes - Shape mismatch: (1, 4) vs (300, 4)
[FAIL] Confs - Shape mismatch: (1,) vs (300,)
[FAIL] Classes - Shape mismatch: (1,) vs (300,)
[WARN] Masks - One is None → Skipping

**Comment by UltralyticsAssistant** [comment]
👋 Hello @myungki-park, thank you for bringing this up and for providing such a detailed description and script! 🚀 This is an automated response to help you get the fastest and most accurate assistance. An Ultralytics engineer will also review your issue and respond shortly.

We recommend referencing the [Docs](https://docs.ultralytics.com/) for extensive [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, which may answer common questions about model exports and inference.

Since you are observing discrepancies between .pt and .engine inference results in YOLO11 segmentation, please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/). This helps us reproduce and debug the issue efficiently. If the script you included is your MRE, that's perfect! If possible, also include a small example image and any relevant outputs or logs that highlight the differences.

Join our community for real-time discussion on [Discord](https://discord.com/invite/ultralytics) 🎧, dive deeper on [Discourse](https://community.ultralytics.com/), or share insights on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Please verify your issue with the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in the following up-to-date verified environments (all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your detailed report and for using Ultralytics! 🦾

**Comment by Y-T-G** [comment]
You shouldn't ignore warnings. They are shown for a reason.

`WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.`

You need to provide the `task` while loading the model.

`model = YOLO("model.engine", task="segment")`

**Comment by myungki-park** [comment]
@Y-T-G yeah you are right. thank you so much!!
